src/main/java/com/dabsquared/gitlabjenkins/connection/GitLabConnection.java

```diff
@@ -53,65 +55,32 @@ public class GitLabConnection extends AbstractDescribableImpl<GitLabConnection>
     private transient String apiToken;
     // TODO make final when migration code gets removed
     private String apiTokenId;
-    private GitLabClientBuilder clientBuilder;
```

Comment:
This hunk should be reverted I think. The existing code is serializing to disk an XML object with an ID (either 'autodetect', 'v3', or 'v4') and an ordinal (integer version). We cannot change this existing format; see https://www.jenkins.io/doc/developer/persistence/backward-compatibility/ for more information.

Yes I understand that v3 support was removed from GitLab 11. At the same time, v5 support will be added in the future. And existing users may have the value set to 'autodetect', which tries each endpoint until it finds one that is working. And I think we want to continue to support this. For example, in the future, GitLab X will have support for v4 and v5, at which point this plugin would want to support that while retaining support for GitLab X-1 which only supports v4. My point is that the autodetect mode is valuable and should not be ripped out.

To keep support for the autodetect mode with the GitLabApi, you'd want to enhance whatever method creates the `GitLabApi` object (in the current PR, `createGitLabApi`) to iterate through all API versions in `org.gitlab4j.api.GitLabApi.ApiVersion` and try each one until finding one that works. For example (in pseudocode, not tested):

```java
import org.gitlab4j.api.GitLabApi.ApiVersion;

    private GitLabApi createGitLabApi(...) {
        for (version : GitLabApi.ApiVersion.versions) {
            GitLabApi gitLabApi =
                    new GitLabApi(version, ...);
            // make sure the server supports this version and loop around otherwise. for example...
            try {
                gitLabApi.getCurrentUser();
                return client;
            } catch (...) {
                // try a lower version
            }
        }   
    }   
```

Yes, to test this you'll need to spin up both a GitLab instance that supports V4 and one that supports V3.

---

Finally there is a _separate_ question of when to remove support for V3 from the Jenkins GitLab plugin (note that the GitLab4J library still supports both V3 and V4). According to the README we no longer promise to support versions of GitLab that are that old, so we could rip out support for V3. That could be done in a _separate_ PR that is targeting the main branch if desired, but it is not related to the current RESTEasy migration. If you want to pursue that unrelated cleanup, please open a separate PR to the main branch that does the following:

- Displays a warning on the configuration page when an unsupported version like V3 is configured (note that to preserve compatibility, we need to maintain V3 as a configuration option forever).
- Displays an error if someone has V3 configured and actually tries to do a build with it.